### ISSUE TITLE
Records not linking to county boundaries [was "Download problem"]

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#81]
Steve Hannah at the TWIC records centre has been having problems with downloads not working for some time now. I had hoped that the recent upgrades would have resolved this, but he's been in touch again:
"It had looked as if it resolved; however, when I went to use it today, I attempted to download from one survey, the records for one month and I’m getting blank spreadsheets.  Same if I try other surveys and date ranges."

Is there a fix available for this?

**Comment by johnvanbreda** [comment]
I checked recent records and they seemed OK, so I think this relates to the old problem we discussed in June with the TWIC boundary having been updated to a broken one. Although I've managed to fix the boundary so new records are indexed OK, the query required to identify the indexation needed to fill the missing gaps is crashing on warehouse1. We are therefore waiting for Biren to upgrade the testwarehouse PostgreSQL to a later version so we can do further testing on it, before trying again on warehouse1.

**Comment by BirenRathod** [comment]
I have now upgraded PostgreSQL to version 9.5.4. on test warehouse.

**Comment by kitenetter** [comment]
Are we able to proceed to testing this?

**Comment by johnvanbreda** [comment]
Now fixed.
Although v9.5 of PostgreSQL prevented the crash, the query took far to long (>12 hrs). I looked at various ways of optimising the query and eventually came up with a way that runs in a few minutes. I've managed to run this OK on the live warehouse so all the TWIC samples are now indexed properly. For reference here are the queries:
```
-- find all that might clip the boundary, using bounding box
select s.id, s.geom
into temporary tempsamples 
from samples s, locations l
where st_intersects(st_envelope(l.boundary_geom), s.geom)
and l.id=1429;

delete from tempsamples where id in (select sample_id from index_locations_samples);

CREATE INDEX ix_tempsamples_geom
  ON tempsamples
  USING gist
  (geom);

insert into index_locations_samples(location_id, sample_id, contains, location_type_id)
select distinct l.id, s.id, st_contains(l.boundary_geom, s.geom), l.location_type_id
    from locations l
    join tempsamples s on (st_intersects(l.boundary_geom, s.geom) and not st_touches(l.boundary_geom, s.geom))
    left join index_locations_samples ils on ils.location_id=l.id and ils.sample_id=s.id
    where ils.id is null
    and l.deleted=false
    and (l.id in (1429))
```

**Comment by kitenetter** [comment]
Have heard from TWIC that they are still experiencing difficulties with downloading data. For example, if they use their LRC download page to download all the data for the "iRecord mammals" survey, they get 19 records (I can replicate that if I masquerade as "TWIC Admin"). If I use the TWIC boundary as a filter in "All records" I can also only find 19 iRecord mammals records. But if I draw an approximate freehand boundary for the TWIC region there are about 400 iRecord mammals records. 

Is this likely to be a continuing problem with the boundary, or could there be some other issue preventing the download?

**Comment by johnvanbreda** [comment]
Should be ffixed now. There was an error in the script - the delete statement should have been:
```
delete from tempsamples where id in (select sample_id from index_locations_samples where location_id=1429);
```

**Comment by kitenetter** [comment]
That has increased the numbers of records available within the TWIC boundary, but still doesn't seem to have sorted out all of them. If I masquerade as TWIC and go to Explore records, and then apply filters to show only the records from source "iRecord mammals", and to use the TWIC boundary from the LRC list, I get this:

![irecord mammals with twic boundary](https://cloud.githubusercontent.com/assets/16039111/18629827/27ad959c-7e60-11e6-9f5f-7e7c1c92e0fb.PNG)

If I then remove the TWIC boundary and look at all the records from source "iRecord mammals" I get this, which seems to have quite a lot of additional records for the north-west part of the TWIC area (e.g. for Loch Lomond/Trossachs):

![irecord mammals without twic boundary](https://cloud.githubusercontent.com/assets/16039111/18629849/5bc66778-7e60-11e6-8363-4e33ef1170bc.PNG)

I haven't checked whether this is also the case with the downloaded data.

**Comment by johnvanbreda** [comment]
Hi @kitenetter 
If you draw the map for TWIC, then compare with a map for the Central Scotland Wildlife Information Service you will see there is a hefty overlap in the region of the Trossachs. Is this correct? Basically the code will only index the records to 1 LRC boundary so the overlap means one of them will miss out.
Regards
John

**Comment by kitenetter** [comment]
@johnvanbreda I'm fairly sure that Central Scotland Wildlife Information Service no longer exists, or at least has been incorporated into TWIC, but am seeking confirmation.

**Comment by kitenetter** [comment]
New boundaries have now been provided by ALERC, will discuss implementation with @BirenRathod

**Comment by BirenRathod** [comment]
Hello Martin
There are some problems in four counties in new boundaries. Also, there are some 10 counties' boundaries are either missing or not properly included.

I have emailed you in detail about this.

**Comment by DavidHepper** [comment]
Martin, may I suggest that the original title of a post such as this is both made more explicit than "Download problem" and, if necessary, updated after investigation has found the cause? This would make it easier to find relevant issues.

**Comment by kitenetter** [comment]
LERC boundary issues now being addressed under #64 - closing.
